File: java/src/main/java/com/google/fhir/common/ResourceUtils.java
//    Copyright 2018 Google Inc.
//
//    Licensed under the Apache License, Version 2.0 (the "License");
//    you may not use this file except in compliance with the License.
//    You may obtain a copy of the License at
//
//        https://www.apache.org/licenses/LICENSE-2.0
//
//    Unless required by applicable law or agreed to in writing, software
//    distributed under the License is distributed on an "AS IS" BASIS,
//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//    See the License for the specific language governing permissions and
//    limitations under the License.

package com.google.fhir.common;

import com.google.common.base.CaseFormat;
import com.google.common.base.Splitter;
import com.google.fhir.r4.core.Bundle;
import com.google.fhir.r4.core.Id;
import com.google.fhir.r4.core.ReferenceId;
import com.google.fhir.wrappers.IdWrapper;
import com.google.protobuf.Descriptors.Descriptor;
import com.google.protobuf.Descriptors.FieldDescriptor;
import com.google.protobuf.Message;
import java.util.ArrayList;
import java.util.Collection;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/** Helper methods for handling FHIR resource protos. */
public final class ResourceUtils {

  public static String getResourceType(Message message) {
    if (!AnnotationUtils.isResource(message)) {
      throw new IllegalArgumentException(
          "Message type " + message.getDescriptorForType().getFullName() + " is not a resource.");
    }
    return message.getDescriptorForType().getName();
  }

  public static String getResourceId(Message message) {
    FieldDescriptor id = message.getDescriptorForType().findFieldByName("id");
    if (!AnnotationUtils.isResource(message) || id == null) {
      throw new IllegalArgumentException(
          "Message type " + message.getDescriptorForType().getFullName() + " is not a resource.");
    }
    if (!message.hasField(id)) {
      return null;
    }
    return ((Id) message.getField(id)).getValue();
  }

  private static Message getContainedResourceInternal(Message resource) {
    Collection<Object> values = resource.getAllFields().values();
    if (values.size() == 1) {
      return (Message) values.iterator().next();
    } else {
      return null;
    }
  }

  public static Message getContainedResource(
      com.google.fhir.stu3.proto.ContainedResource resource) {
    return getContainedResourceInternal(resource);
  }

  public static Message getContainedResource(com.google.fhir.r4.core.ContainedResource resource) {
    return getContainedResourceInternal(resource);
  }

  @SuppressWarnings("unchecked")
  public static <V> V getValue(Message primitive) {
    return (V) primitive.getField(primitive.getDescriptorForType().findFieldByName("value"));
  }

  /*
   * Convert any absolute references in the provided bundle to relative references, assuming the
   * targets of the references are also present in the bundle.
   */
  public static Bundle resolveBundleReferences(Bundle bundle) {
    Map<String, String> referenceMap = new HashMap<>();
    for (Bundle.Entry entry : bundle.getEntryList()) {
      if (entry.hasFullUrl()) {
        Message resource = ResourceUtils.getContainedResource(entry.getResource());
        String resourceType = ResourceUtils.getResourceType(resource);
        String resourceId = ResourceUtils.getResourceId(resource);
        String relativeReference = resourceType + "/" + resourceId;
        referenceMap.put(entry.getFullUrl().getValue(), relativeReference);
      }
    }
    return (Bundle) replaceReferences(bundle, referenceMap);
  }

  private static Message replaceOneReference(Message message, Map<String, String> referenceMap) {
    FieldDescriptor uri = message.getDescriptorForType().findFieldByName("uri");
    if (!message.hasField(uri)) {
      return message;
    }
    String oldValue = ((com.google.fhir.r4.core.String) message.getField(uri)).getValue();
    if (!referenceMap.containsKey(oldValue)) {
      return message;
    }
    Message.Builder builder = message.toBuilder();
    String newValue = referenceMap.get(oldValue);
    builder.setField(uri, com.google.fhir.r4.core.String.newBuilder().setValue(newValue).build());
    return splitIfRelativeReference(builder);
  }

  @SuppressWarnings("unchecked")
  private static Message replaceReferences(Message message, Map<String, String> referenceMap) {
    Message.Builder builder = null;
    for (Map.Entry<FieldDescriptor, Object> field : message.getAllFields().entrySet()) {
      if (field.getKey().getType() == FieldDescriptor.Type.MESSAGE) {
        Object newValue = field.getValue();
        if (field.getKey().isRepeated()) {
          List<Message> newList = new ArrayList<>();
          for (Message item : (List<Message>) field.getValue()) {
            Message newItem;
            if (AnnotationUtils.isReference(item)) {
              newItem = replaceOneReference(item, referenceMap);
            } else {
              newItem = replaceReferences(item, referenceMap);
            }
            if (newItem != item) {
              newValue = newList;
            }
            newList.add(newItem);
          }
        } else {
          Message value = (Message) field.getValue();
          if (AnnotationUtils.isReference(field.getKey().getMessageType())) {
            newValue = replaceOneReference(value, referenceMap);
          } else {
            newValue = replaceReferences(value, referenceMap);
          }
        }
        if (newValue != field.getValue()) {
          if (builder == null) {
            builder = message.toBuilder();
          }
          builder.setField(field.getKey(), newValue);
        }
      }
    }
    if (builder == null) {
      return message;
    } else {
      return builder.build();
    }
  }

  /*
   * Split relative references into their components, for example, "Patient/ABCD" will result in
   * the patientId field getting the value "ABCD".
   */
  public static Message splitIfRelativeReference(Message.Builder builder) {
    Descriptor descriptor = builder.getDescriptorForType();
    FieldDescriptor uriField = descriptor.findFieldByName("uri");
    if (!builder.hasField(uriField)) {
      return builder.build();
    }
    Message uri = (Message) builder.getField(uriField);
    String uriValue = (String) uri.getField(uri.getDescriptorForType().findFieldByName("value"));
    if (uriValue.startsWith("#")) {
      FieldDescriptor fragmentField = descriptor.findFieldByName("fragment");
      Message.Builder fragmentBuilder = builder.getFieldBuilder(fragmentField);
      return ProtoUtils.fieldWiseCopy(
              com.google.fhir.r4.core.String.newBuilder()
                  .setValue(new IdWrapper(uriValue.substring(1)).getWrapped().getValue()),
              fragmentBuilder)
          .build();
    }
    // Look for references of type "ResourceType/ResourceId"
    List<String> parts = Splitter.on('/').splitToList(uriValue);
    if (parts.size() == 2 || (parts.size() == 4 && "_history".equals(parts.get(2)))) {
      String resourceFieldName =
          CaseFormat.UPPER_CAMEL.to(CaseFormat.LOWER_UNDERSCORE, parts.get(0)) + "_id";
      FieldDescriptor field = descriptor.findFieldByName(resourceFieldName);
      if (field == null) {
        throw new IllegalArgumentException(
            "Invalid resource type in reference: " + resourceFieldName + ":" + parts.get(0));
      } else {
        ReferenceId.Builder refId =
            ReferenceId.newBuilder().setValue(new IdWrapper(parts.get(1)).getWrapped().getValue());
        if (parts.size() == 4) {
          refId.setHistory(new IdWrapper(parts.get(3)).getWrapped());
        }
        ProtoUtils.fieldWiseCopy(refId, builder.getFieldBuilder(field));
        return builder.build();
      }
    }
    // Keep the uri field.
    return builder.build();
  }
}

File: py/google/fhir/core/utils/resource_utils.py
#
# Copyright 2021 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""A collection of functions for dealing with FHIR resources."""

from typing import Any, List, Type, TypeVar, cast

from google.protobuf import message
from google.fhir.core.utils import fhir_types
from google.fhir.core.utils import path_utils
from google.fhir.core.utils import proto_utils

_T = TypeVar('_T', bound=message.Message)


def extract_resources_from_bundle(bundle: message.Message, *,
                                  resource_type: Type[_T]) -> List[_T]:
  """Returns a list of resources of type `resource_type` from `bundle`.

  Args:
    bundle: The FHIR Bundle to examine.
    resource_type: The message type of the resource to return.

  Returns:
    A list of resources of type `resource_type` belonging to the bundle.

  Raises:
    TypeError: In the event that `bundle` is not of type "Bundle".
    ValueError: In the event that a field corresponding to the "snake_case" name
    of `resource_type` does not exist on `Bundle.Entry`.
  """
  # TODO(b/148949073): Replace with templated/consolidated `fhir_types` call.
  if not fhir_types.is_type_or_profile_of(
      'http://hl7.org/fhir/StructureDefinition/Bundle', bundle):
    raise TypeError(
        f'{bundle.DESCRIPTOR.name} is not a type or profile of Bundle.')

  contained_resource_field = path_utils.camel_case_to_snake_case(
      resource_type.DESCRIPTOR.name)
  return [
      getattr(entry.resource, contained_resource_field)
      for entry in cast(Any, bundle).entry
      if entry.resource.HasField(contained_resource_field)
  ]


def get_contained_resource(
    contained_resource: message.Message) -> message.Message:
  """Returns the resource instance contained within `contained_resource`.

  Args:
    contained_resource: The containing `ContainedResource` instance.

  Returns:
    The resource contained by `contained_resource`.

  Raises:
    TypeError: In the event that `contained_resource` is not of type
    `ContainedResource`.
    ValueError: In the event that the oneof on `contained_resource` is not set.
  """
  # TODO(b/154059162): Use an annotation here.
  if contained_resource.DESCRIPTOR.name != 'ContainedResource':
    raise TypeError('Expected `ContainedResource` but got: '
                    f'{type(contained_resource)}.')
  oneof_field = contained_resource.WhichOneof('oneof_resource')
  if oneof_field is None:
    raise ValueError('`ContainedResource` oneof not set.')
  return proto_utils.get_value_at_field(contained_resource, oneof_field)
